### DISCUSSION TITLE
Splitting tests files in smaller files in components/modules tests

### DISCUSSION THREAD

**Discussion by esciara** [home-assistant/architecture#1008] [General]
## Context

I was [told on Discord's dev channel](https://discord.com/channels/330944238910963714/554842238073700352/1182642436640608257) that this was probably the place to put my question/proposal.

One thing that surprised me when I first tried to understand HA code through the testing code, is that some `test_*.py` files (172 actually) have thousands lines of code (I even saw some with 5000+ lines). This makes it ... so ... hard to go through it because there is no visible organisation or grouping of what part of the component you test at what stage. Let alone the fact that it puts my IDE on its knees... Where I come from, we make it nearly mandatory to group tests in smaller files, grouping them by areas or even sub areas they are testing, which makes it much easier to navigate.

I proposed the PR https://github.com/home-assistant/core/pull/105036 to work on it but the answer was that it was not the standard and that tests are grouped by module, not by functionality. I might have interpreted that as a definite "no" a bit quickly, without offering an alternative (even though I am a newbie and I am conscious I don't have any legitimacy around HA, at least not yet 😅 😜) .

So just wanted to check here and make another proposal before I bother anyone asking to review a modified PR.

## Proposal

Would it be ok/a good idea to split the tests, but this time grouping them by modules as a directory with tests files in it?

Instead of:

```
home-assistant-core/tests/components/generic_thermostat
├── __init__.py
├── fixtures
│   └── configuration.yaml
└── test_climate.py         <= 1600+ lines = 🙀
```

... We would end up with something along the lines of: 

```
home-assistant-core/tests/components/generic_thermostat
├── __init__.py
├── climate
│   ├── conftest.py
│   ├── const.py
│   ├── common.py
│   ├── test_component_reload.py
│   ├── test_component_restore_state.py
│   ├── test_component_setup.py
│   ├── test_cooling_hvac_mode.py
│   ├── test_cooling_keep_alive.py
│   ├── test_cooling_min_cycle_duration.py
│   ├── test_cooling_switch.py
│   ├── test_current_temperature.py
│   ├── test_heating_hvac_mode.py
│   ├── test_heating_keep_alive.py
│   ├── test_heating_min_cycle_duration.py
│   ├── test_heating_preset_mode.py
│   ├── test_heating_switch.py
│   └── test_precision.py
└── fixtures
    └── configuration.yaml
```

## Consequences

- The parts of the module being tested become instantaneously clearer.
- Much easier to navigate and understand, especially for newcomers who want to participate. This decreases the price of the entry ticket.
- IDEs are more responsive and support feature appear faster
- Tests outputs are better compartimentized and one would immediately see which set of tests break (did not show breaking tests in the screenshots, but I guess you would get the point):
  - Tests through the IDE UI (old vs new):  
     <img width="496" alt="image" src="https://github.com/home-assistant/architecture/assets/1336689/4f3a577f-2808-4714-9541-33ba52723fb3">    
     <img width="496" alt="image" src="https://github.com/home-assistant/architecture/assets/1336689/f174e9b8-5dca-4c6b-b2c3-1035a6d4145a">  
   - Tests through cli (old vs new):
     ![image](https://github.com/home-assistant/architecture/assets/1336689/c7b500f6-ad1a-45d6-bc75-33e9c6c674fa)  
     <img width="894" alt="image" src="https://github.com/home-assistant/architecture/assets/1336689/8b6e5952-84a4-49a6-8689-03b0686eb0d8">

**Comment by rytilahti** [comment]
Yeah, this is definitely a better place for such discussions than the discord channels :-)

Personally, I am against forcing splitting up the tests like this. Not only as it might cause friction when deciding into which file a new test method should go in, but finding a test would require using an IDE (which might be slow, given the size of the code base) or doing some grepping in the test directory. ctrl+f is just more comfortable when you know what you are looking for.

As not all components are equal, maybe the decision on how to organize the test code (similarly how it's done for the implementation) should be left to the codeowners? While most of the components follow a similar structure, there are some (e.g., mqtt, sonos) that have their own style with more than a single file per platform. In fact, sonos even has a separate `test_plex_playback.py` with no similarly named module in its integration directory.

That being said, if the goal is to make the tests more approachable for newcomers, I'd rather suggest to have them stored inside the integration directory to keep them "local". Having all the relevant files nearby in the directory hierarchy makes them directly visible for anyone who's reading the code without jumping around the code base.

**Reply by Jc2k** [reply]
For me, the current approach scales to meet my needs, it's easy to know where my tests belong. Even though when I started I was a HA noob I didn't really feel like I had any problems picking up what HA did. So I don't really want to change.

**Reply by esciara** [reply]
> Personally, I am against forcing splitting up the tests like this. Not only as it might cause friction when deciding into which file a new test method should go in, but finding a test would require using an IDE (which might be slow, given the size of the code base) or doing some grepping in the test directory.

Hi. I was by no means suggesting that this should be enforced, just allowed (or in my wildest dreams recommended?) as a possible way to organise the tests. 

> ctrl+f is just more comfortable when you know what you are looking for.

That is one of my points: you know what you are looking for when you have spent a significant amount of time digging in the code and figuring out how it all works, which I found very difficult as a new comer (and I consider myself a seasoned Python developer). From an outsider's perspective who would like to participate and needs to fit to grip with the concepts and the code base, I find it quite daunting. 

I am not saying that it should be all bells and whistles from day one and that no effort should be required to learn how to add/correct things on HA. I am just saying it is much, much harder than it could/should be. 

> As not all components are equal, maybe the decision on how to organize the test code (similarly how it's done for the implementation) should be left to the codeowners? While most of the components follow a similar structure, there are some (e.g., mqtt, sonos) that have their own style with more than a single file per platform. In fact, sonos even has a separate `test_plex_playback.py` with no similarly named module in its integration directory.

Absolutely. As I said, recommendations as possible way to organise tests. 

> That being said, if the goal is to make the tests more approachable for newcomers, I'd rather suggest to have them stored inside the integration directory to keep them "local". Having all the relevant files nearby in the directory hierarchy makes them directly visible for anyone who's reading the code without jumping around the code base.

Do you mean having a structure like in `go` or as I saw in some `javascript` projects, like this?:


```
home-assistant-core/home-assistant/components/generic_thermostat
├── __init__.py
├── climate.py
├── test_climate.py
```

Personally, I find this soooo un-pythonic... (actually I hate it, but that is personal choice/opinion). 

My 2 cents.

**Comment by frenck** [comment]
I do have some personal opinions on this one as well. That said, the proposal is clear. I will take this into our weekly meeting to sync up. Will get back to this one once it has been discussed.

../Frenck

**Comment by allenporter** [comment]
I would be worried this introduces additional arbitrary file naming conventio.. While this may make one test file easier to understand it will make the code base harder to navigate as a whole and also lead to things like duplicate tests or test files with unclear scope.  For example, I have a bug when changing from heating to cooling: Should is stick this in "test_cooling_hvac_mode.py" or "test_heating_hvac_mode.py"?  Additional discussions where people are debating naming will be distracting. I'm not sure what newcomers find easier to use, so you may be right there, i'm just not sure its worth the other side effects. 

I think generic_thermostat has other issues, and is not following modern design standards. I don't think it is the best example to optimize for necessarily.

Encouraging usage of fixtures shared across tests seems fine, and is fairly typical.

**Reply by esciara** [reply]
> I think generic_thermostat has other issues, and is not following modern design standards. I don't think it is the best example to optimize for necessarily.

I know and this is part of the reason I took this example, alongside the fact this is close to what I try to build. I find refactoring to be one of the best ways to understand code. I would even be quite willing to make this code to follow modern standard, as it would help me understand what modern standard is. And having clean tests that can be read without getting a headache is an open door to that.

> Should is stick this in "test_cooling_hvac_mode.py" or "test_heating_hvac_mode.py"?

Actually that is an interesting question, which to me potentially point out the fact that one might not need to have tests for one and the other. If you go through the same code to handle both cases, then you don't need to have to sets of tests. That would reduce the footprint of your test, in turn make that you have less tests to read and maintain, in turn making the whole thing clearer and simpler to grasp and work with.

As usual: my 2 cents.

**Comment by parautenbach** [comment]
Naming things is famously a hard problem – especially in programming. 

I don't have HA coding credentials, besides for a custom component or so, but I have coded commercially/professionally for a very long time, but also before that and up to this day. 

Structure in code is important – I think everyone would agree on that. The question is how to go about it. 

To me, the structure of the test files should reflect (or be aligned) with the code under testing. If you cannot break up your test files, one probably needs to start with breaking up the code under testing. 

In this specific case, there really is just one module and one class involved, so there's nothing to restructure. Personally, I would keep with exactly how it is now. If there were multiple classes, one could consider a test file per class. 

When it comes to the main code, it is a general best practice to keep files, modules and classes manageable, but tests are different. Test files tend to be large and long (but individual tests should not be). 

A better option here might be custom [pytest marks](https://docs.pytest.org/en/7.4.x/how-to/mark.html). I know this is typically used to execute tests, but it might be useful in this case to find similar or related tests quicker. (Plus, it would actually be useful in cases where one wants to run a subset of tests very often as part of the development of a feature or change, before doing full regression.)

In the end, there will always be some ambiguity or duplication. The advantage of marks is that multiple can be applied, as opposed to files where a test can only be in one file.
